x/src/lint/license.rs

```diff
@@ -1,54 +1,80 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-use std::{borrow::Cow, convert, ffi::OsStr, path::Path};
+use crate::lint::engine::prelude::*;
 
 static LICENSE_HEADER: &str = "Copyright (c) The Libra Core Contributors\n\
                                SPDX-License-Identifier: Apache-2.0\n\
                                ";
 
-pub(super) fn has_license_header(file: &Path, contents: &str) -> Result<(), Cow<'static, str>> {
-    enum FileType {
-        Rust,
-        Shell,
-        Proto,
+#[derive(Copy, Clone, Debug)]
+pub(super) struct LicenseHeader;
+
+impl Linter for LicenseHeader {
+    fn name(&self) -> &'static str {
+        "license-header"
     }
+}
 
-    let file_type = match file
-        .extension()
-        .map(OsStr::to_str)
-        .and_then(convert::identity)
-    {
-        Some("rs") => FileType::Rust,
-        Some("sh") => FileType::Shell,
-        Some("proto") => FileType::Proto,
-        _ => return Ok(()),
-    };
-
-    // Determine if the file is missing the license header
-    let missing_header = match file_type {
-        FileType::Rust | FileType::Proto => {
-            let maybe_license = contents
-                .lines()
-                .skip_while(|line| line.is_empty())
-                .take(2)
-                .map(|s| s.trim_start_matches("// "));
-            !LICENSE_HEADER.lines().eq(maybe_license)
+impl ContentLinter for LicenseHeader {
+    fn pre_run<'l>(&self, file_ctx: &FileContext<'l>) -> Result<RunStatus<'l>> {
+        match file_ctx.extension() {
+            Some("rs") | Some("sh") | Some("proto") => Ok(RunStatus::Executed),
+            other => Ok(RunStatus::Skipped(SkipReason::UnsupportedExtension(other))),
         }
-        FileType::Shell => {
-            let maybe_license = contents
-                .lines()
-                .skip_while(|line| line.starts_with("#!"))
-                .skip_while(|line| line.is_empty())
-                .take(2)
-                .map(|s| s.trim_start_matches("# "));
-            !LICENSE_HEADER.lines().eq(maybe_license)
+    }
+
+    fn run<'l>(
+        &self,
+        ctx: &ContentContext<'l>,
+        out: &mut LintFormatter<'l, '_>,
+    ) -> Result<RunStatus<'l>> {
+        enum FileType {
+            Rust,
+            Shell,
+            Proto,
         }
-    };
 
-    if missing_header {
-        return Err("missing a license header".into());
-    }
+        let content = match ctx.content() {
+            Some(content) => content,
+            None => {
+                // This is not a UTF-8 file -- don't analyze it.
+                return Ok(RunStatus::Skipped(SkipReason::NonUtf8));
+            }
+        };
 
-    Ok(())
+        let file_type = match ctx.file_ctx().extension() {
+            Some("rs") => FileType::Rust,
+            Some("sh") => FileType::Shell,
+            Some("proto") => FileType::Proto,
+            _ => unreachable!("filtered out in pre_run"),
+        };
```

Comment:
Thinking about this, it might be a little more robust if the filetype enum was scoped to the whole module and in the pre_run we convert the extension to the enum via something that returned an Option<FileType>, that way we're checking for the extensions in a single place instead of duplicating that logic here and above in the pre_run.

Reply:
Ahh, that would require carrying some state between `pre_run` and `run`. I actually want to allow doing that on a `FileContext` through some sort of type-indexed map (something like [`shred::World`](https://docs.rs/shred/0.10.2/shred/struct.World.html)) but I haven't fully thought this through

Reply:
I can add a `TODO` comment though, thanks!~

Reply:
Oh no I'm not even talking about sharing state between the two. I'm just talking about the logic you're using to check for the file extensions is being duplicated between both the pre_run and run functions. I was just trying to say you could avoid duplicating that logic.

Reply:
Oh I see. Let me do that.